pkg/client/testing/core/fixture.go

```diff
@@ -49,153 +58,158 @@ type ObjectScheme interface {
+		// not on interfaces, as some interfaces are identical
+		// (e.g. UpdateAction and CreateAction), so if we use them,
+		// updates and creates end up matching the same case branch.
+		switch typedAction := action.(type) {
+
+		case ListActionImpl:
 			kind.Kind += "List"
-			resource, err := o.Kind(kind, "")
-			return true, resource, err
+			obj, err := registry.Get(kind, ns, "")
+			return true, obj, err
 
-		case GetAction:
-			resource, err := o.Kind(kind, castAction.GetName())
-			return true, resource, err
+		case GetActionImpl:
+			obj, err := registry.Get(kind, ns, typedAction.GetName())
+			return true, obj, err
 
-		case DeleteAction:
-			resource, err := o.Kind(kind, castAction.GetName())
-			return true, resource, err
+		case CreateActionImpl:
+			objMeta, err := meta.Accessor(typedAction.GetObject())
+			if err != nil {
+				return true, nil, err
+			}
+			if typedAction.GetSubresource() == "" {
+				err = registry.Add(typedAction.GetObject())
+			} else {
+				// Handle subresource creation as an update on the
+				// enclosing resource.
+				err = registry.Update(typedAction.GetObject())
+			}
+			if err != nil {
+				return true, nil, err
+			}
+			obj, err := registry.Get(kind, ns, objMeta.GetName())
+			return true, obj, err
 
-		case CreateAction:
-			accessor, err := meta.Accessor(castAction.GetObject())
+		case UpdateActionImpl:
+			objMeta, err := meta.Accessor(typedAction.GetObject())
+			if err != nil {
+				return true, nil, err
+			}
+			err = registry.Update(typedAction.GetObject())
 			if err != nil {
 				return true, nil, err
 			}
-			resource, err := o.Kind(kind, accessor.GetName())
-			return true, resource, err
+			obj, err := registry.Get(kind, ns, objMeta.GetName())
+			return true, obj, err
 
-		case UpdateAction:
-			accessor, err := meta.Accessor(castAction.GetObject())
+		case DeleteActionImpl:
+			err := registry.Delete(kind, ns, typedAction.GetName())
 			if err != nil {
 				return true, nil, err
 			}
-			resource, err := o.Kind(kind, accessor.GetName())
-			return true, resource, err
+			return true, nil, nil
 
 		default:
 			return false, nil, fmt.Errorf("no reaction implemented for %s", action)
 		}
 	}
 }
 
-// AddObjectsFromPath loads the JSON or YAML file containing Kubernetes API resources
-// and adds them to the provided ObjectRetriever.
-func AddObjectsFromPath(path string, o ObjectRetriever, decoder runtime.Decoder) error {
-	data, err := ioutil.ReadFile(path)
-	if err != nil {
-		return err
-	}
-	data, err = yaml.ToJSON(data)
-	if err != nil {
-		return err
-	}
-	obj, err := runtime.Decode(decoder, data)
-	if err != nil {
-		return err
-	}
-	if err := o.Add(obj); err != nil {
-		return err
-	}
-	return nil
-}
-
-type objects struct {
-	types   map[string][]runtime.Object
-	last    map[string]int
+type registry struct {
 	scheme  ObjectScheme
 	decoder runtime.Decoder
+	mu      sync.RWMutex
+	objects map[string][]runtime.Object
```

Comment:
> Also, when setting tests up, objects are being added to the clientset directly, at that point we don't know what GroupVersionResource to associate them with, since we don't have any REST context at that point, just a runtime.Object.

Right. Perhaps we can ask the caller to provide GVR along with the objects, but that will be a separate PR.